parquet/src/main/java/org/apache/iceberg/parquet/ParquetVariantUtil.java

```diff
@@ -403,19 +400,14 @@ public Type array(VariantArray array, List<Type> elementResults) {
         return null;
       }
 
-      // Choose most common type as shredding type and build 3-level list
-      Type defaultTYpe = elementResults.get(0);
-      Type shredType =
-          elementResults.stream()
-              .filter(Objects::nonNull)
-              .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()))
-              .entrySet()
-              .stream()
-              .max(Map.Entry.comparingByValue())
-              .map(Map.Entry::getKey)
-              .orElse(defaultTYpe);
-
-      return list(shredType);
+      // Shred if all the elements are of a uniform type and build 3-level list
+      Type shredType = elementResults.get(0);
+      if (shredType != null
+          && elementResults.stream().allMatch(type -> Objects.equals(type, shredType))) {
+        return list(shredType);
+      }
+
+      return null;
```

Comment:
Do you think there are cases where we want to shred an array of mixed types, with each value stored in `value` as an encoded variant? I'm not sure.

Reply:
I don't think we will have the case that all the elements will be stored in `value` since I would expect engines will at least choose a shredding type that matches some of the elements. I think this is more common: by sampling the elements, we choose INT32 as the shredding type while some are strings. If all of theme are in `value` and they are binary, that doesn't seem to be useful.